src/vicsetup/libvicsetup/dm.c

```diff
@@ -0,0 +1,917 @@
+            num_blocks,
+            hash_offset,
+            hash_alg,
+            root_digest_ascii,
+            salt_ascii);
+
+        if (n <= 0 || (size_t)n >= sizeof(params))
+            RAISE(VIC_BUFFER_TOO_SMALL);
+    }
+
+    /* Create the device-mapper task object */
+    if (!(dmt = dm_task_create(DM_DEVICE_CREATE)))
+        RAISE(VIC_FAILED);
+
+    /* Set the device-mapper name */
+    if (!dm_task_set_name(dmt, dm_name))
+        RAISE(VIC_FAILED);
+
+    if (!dm_task_secure_data(dmt))
+        RAISE(VIC_FAILED);
+
+    if (!dm_task_set_ro(dmt))
+        RAISE(VIC_FAILED);
+
+#ifdef TRACE_TARGET
+    printf("DM-NAME: %s\n", dm_name);
+    printf("TARGET: start{%lu} size{%lu} target{%s} params{%s}\n",
+        start, size, target, params);
+#endif
+
+    /* Set the target */
+    if (!dm_task_add_target(dmt, start, size, target, params))
+        RAISE(VIC_FAILED);
+
+    /* Run the task to create the new target type */
+    if (!dm_task_run(dmt))
+        RAISE(VIC_FAILED);
+
+    /* Verify that the target now exists */
+    {
+        struct dm_info dmi;
+
+        if (!dm_task_get_info(dmt, &dmi) || !dmi.exists)
+            RAISE(VIC_FAILED);
+    }
+
+done:
+
+    if (root_digest_ascii)
+        vic_free(root_digest_ascii);
+
+    if (salt_ascii)
+        vic_free(salt_ascii);
+
+    if (dmt)
+    {
+        dm_task_destroy(dmt);
+        dm_task_update_nodes();
+    }
+
+    dm_lib_release();
+
+    return result;
+}
+
+vic_result_t vic_dm_remove(const char* name)
+{
+    vic_result_t result = VIC_OK;
+    struct dm_task* dmt = NULL;
+    const size_t max_retries = 10;
+#ifdef USE_UDEV
+    int sync;
+    uint32_t cookie = 0;
+#endif
+
+    if (!name)
+        RAISE(VIC_BAD_PARAMETER);
+
+    if (!(dmt = dm_task_create(DM_DEVICE_REMOVE)))
+        RAISE(VIC_FAILED);
+
+    if (!dm_task_set_name(dmt, name))
+        RAISE(VIC_FAILED);
+
+#ifdef USE_UDEV
+    /* Determine whether sync support is available */
+    if ((sync = dm_udev_get_sync_support()))
+    {
+        uint16_t udev_flags = 0;
+        udev_flags += DM_UDEV_DISABLE_LIBRARY_FALLBACK;
+        dm_task_set_cookie(dmt, &cookie, udev_flags);
+    }
+#endif
+
+    /* Wait until the device is not busy (wait 1 second at most) */
+    for (size_t i = 0; i < max_retries; i++)
+    {
+        if (dm_task_run(dmt))
+        {
+            struct timespec req;
+            const uint64_t second = 1000000000;
+
+            /* Sleep for 1/10th of a second */
+            req.tv_sec = 0;
+            req.tv_nsec = second / 10;
+            nanosleep(&req, NULL);
+            break;
+        }
+    }
+#ifdef USE_UDEV
+    if (sync)
+        dm_udev_wait(cookie);
+#endif
+
+done:
+
+    if (dmt)
+    {
+        dm_task_destroy(dmt);
+        dm_task_update_nodes();
+    }
+
+    dm_lib_release();
+
+    return result;
+}
+
+/*
+**==============================================================================
+**
+** Experimental raw dm-ioctl code (to bypass libdevmapper).
```

Comment:
What is the status of this? Can this be used? What is missing?

Reply:
I removed this experimental code for now. We could spend three days and complete the replacement but I don't see why we would do it now. I originally implemented this to figure out why the kernel was crashing when calling dm-ioctl() to format an integrity device. It turns out that kernel memory was exhausted by the dm-integrity module, which consumes massive amounts of memory when formatting integrity headers (> 64MB). Expanding kernel memory worked around this for now, but we might want to perform integrity formatting in user-space (rather than letting the dm-integrity kernel module do it). I captured this in the following issue: https://github.com/lsds/sgx-lkl/issues/608.